### PR TITLE
Add validate interface in configmodels to force each component do configuration validation

### PR DESCRIPTION
**Pull request by mxiamxia**
**Description:** 
The goal is to add `Validate()` interface in `config.Receiver|Exporter|Processor|Extension` so we can force each component to add validation logic on component level configuration. Only have the changes for `config.Receiver` in this PR.

Added `CustomConfigOptions` interface in `config.Receiver`.  And `CustomConfigOptions` interface has embedded `CustomConfigValidator` interface so it can force each receiver component to implement `Validate()` for Config validation check. Also, `CustomConfigOptions` interface can be extended to add more CustomConfig options.

**The reason why we should have `Validate()` for each component.**
Currently, since we don't have a unified fashion for components to `validate` the configuration. The component contributors write the config validate logic in different places in the component code. 
For example. in `prometheusreceiver` the cfg validate logic is coded in [createMetricsReceiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/prometheusreceiver/factory.go#L105). and in `awsemfexporter` it was written in [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/awsemfexporter/config.go#L100). 
By doing the current implementation, all the contributors will follow the same fashion for the configuration validation. And we can fail the collector in the start time with friendly error messages.

```
type Receiver interface {
	NamedEntity
	CustomConfigOptions
}
// CustomConfigOptions defines the interfaces for configuration customization options
// Different custom interfaces can be added like CustomConfigValidator, CustomConfigUnmarshaller, etc.
type CustomConfigOptions interface {
	CustomConfigValidator
}
// CustomConfigValidator defines the interface for the custom configuration validation on each component
type CustomConfigValidator interface {
	Validate() error
}
```
Custom Validation Example for `prometheusreceiver`,
```
// Validate implements the custom validation check on configuration
func (cfg *Config) Validate() error {
	if cfg.PrometheusConfig == nil || len(cfg.PrometheusConfig.ScrapeConfigs) == 0 {
		return errNilScrapeConfig
	}
	return nil
}
```

Need help to review this approach before moving to the others (exporter, processor, extension)

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-collector/issues/2541
https://github.com/open-telemetry/opentelemetry-collector/issues/2597

**Testing:**
WIP

**Documentation:** 
N/A

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -560,6 +577,14 @@ func defaultUnmarshaler(componentViperSection *viper.Viper, intoCfg interface{})
 	return componentViperSection.UnmarshalExact(intoCfg)
 }
 
+func validator(config configmodels.Receiver) (configmodels.CustomValidator, error) {
+	cfg, ok := config.(configmodels.ConfigValidator)
+	if !ok {
+		return nil, errors.New("validate interface is not implemented by Config")
+	}
+	return cfg.Validate, nil
+}
```
Comment (mx-psi) [inline config/config.go:586]: Why don't we make the `ConfigValidator` interface optional and do here something similar to what we do for unmarshaling?
https://github.com/open-telemetry/opentelemetry-collector/blob/1b164905a92c4c94ece83f66c58250f0531a2be1/config/config.go#L569-L574

We can have a `defaultUnmarshaler` that just does `return nil`.

File: config/configmodels/receiver.go [hunk 2]
```diff
@@ -20,6 +20,24 @@ type Receiver interface {
...
+	ConfigValidator
+	// CustomConfigUnmarshaller
+}
+
+// ConfigValidator defines the interface for the custom configuration validation on each component
+type ConfigValidator interface {
+	Validate() error
+}
+
+// CustomValidator is a function that run the customized validation
```
Comment (mx-psi) [inline config/configmodels/receiver.go:35]:
```suggestion
// CustomValidator is a function that runs the customized validation
```

File: config/configmodels/receiver.go [hunk 3]
```diff
@@ -20,6 +20,24 @@ type Receiver interface {
...
+// on component level configuration.
+// intoCfg interface{}
+//   An empty interface wrapping a pointer to the config struct.
```
Comment (mx-psi) [inline config/configmodels/receiver.go:38]:
```suggestion
```

File: receiver/opencensusreceiver/config.go [hunk 4]
```diff
@@ -33,13 +33,13 @@ type Config struct {
...
+func (cfg *Config) buildOptions() ([]ocOption, error) {
 	var opts []ocOption
-	if len(rOpts.CorsOrigins) > 0 {
-		opts = append(opts, withCorsOrigins(rOpts.CorsOrigins))
+	if len(cfg.CorsOrigins) > 0 {
+		opts = append(opts, withCorsOrigins(cfg.CorsOrigins))
 	}
 
-	grpcServerOptions, err := rOpts.GRPCServerSettings.ToServerOption()
+	grpcServerOptions, err := cfg.GRPCServerSettings.ToServerOption()
```
Comment (mx-psi) [inline receiver/opencensusreceiver/config.go:42]: Not sure this change belongs to this PR

File: config/config.go (hunk 1 above)
Comment (bogdandrutu) [reply config/config.go:586]: We discussed this, and we believe that for validate we want users to be forced to implement it, so they are forced to think about that. Also most of the components implementation that I've seen require some validation which sometimes happens in constructor or sometimes in start (very inconsistent).

For the "unmarshaler" that is a very edge case so going with a separate interface makes sense in that case. Does it make sense?

Comment (mx-psi) [reply config/config.go:586]: Makes sense, yes, thanks! I wasn't aware of that discussion but I agree with your points

Comment (bogdandrutu) [reply config/config.go:586]: I think was an ad-hoc meeting where we discussed that, sorry for not sharing it. Maybe @mxiamxia you can add this to the PR description and explain why Validate is a mandatory thing to have.

**Review by bogdandrutu — COMMENTED** [review]
The way how I see this, based on the discussion in the other comment, I think we should have (in current configmodels):

```golang
package configmodels

// Processor is the configuration of a processor. Specific processors must implement this
// interface and will typically embed ProcessorSettings struct or a struct that extends it.
type Processor interface {
	NamedEntity

	Validate() errror  // and for all interfaces Exporter, etc.
}

// In the unmarshaler PR, optional interface for configs
type Unmarshaler interface {
	Unmarshal(componentParser *config.Parser, intoCfg interface{}) error
}

type UnmarshalFunc func(componentParser *config.Parser, intoCfg interface{}) error
```

File: config/config.go [hunk 5]
```diff
@@ -302,6 +309,16 @@ func LoadReceiver(componentConfig *viper.Viper, typeStr configmodels.Type, fullN
 		return nil, errorUnmarshalError(receiversKeyName, fullName, err)
 	}
 
+	// run the custom configuration validator for the current component
+	// if no custom config validation logic implemented it calls configmodels.Validate
+	validate, err := validator(receiverCfg)
```
Comment (bogdandrutu) [inline config/config.go:314]: We should not validate configs during loading. We should call validate in `configmodels.Config.Validate()` and validate all configs there.

https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configmodels/config.go#L57

File: config/configmodels/receiver.go [hunk 6]
```diff
@@ -20,6 +20,24 @@ type Receiver interface {
 	NamedEntity
 }
 
+// CustomConfigOptions defines the interfaces for configuration customization options
+// Different custom interfaces can be added like CustomConfigValidator, CustomConfigUnmarshaller, etc.
+type CustomConfigOptions interface {
```
Comment (bogdandrutu) [inline config/configmodels/receiver.go:25]: As mentioned in the comment, would be good to make this a `required` func to implement.

File: config/config.go (hunk 1 above)
Comment (mxiamxia) [reply config/config.go]: Explained why should we have a mandatory `Validate()` in the PR description. Thx!

File: config/configmodels/receiver.go (hunk 2 above)
Comment (mxiamxia) [reply config/configmodels/receiver.go]: done

File: receiver/opencensusreceiver/config.go (hunk 4 above)
Comment (mxiamxia) [reply receiver/opencensusreceiver/config.go:42]: golint requires the same receiver name. I changed it to `cfg`

**Comment by mxiamxia** [comment]
> The way how I see this, based on the discussion in the other comment, I think we should have (in current configmodels):
> 
> ```go
> package configmodels
> 
> // Processor is the configuration of a processor. Specific processors must implement this
> // interface and will typically embed ProcessorSettings struct or a struct that extends it.
> type Processor interface {
> 	NamedEntity
> 
> 	Validate() errror  // and for all interfaces Exporter, etc.
> }
> 
> // In the unmarshaler PR, optional interface for configs
> type Unmarshaler interface {
> 	Unmarshal(componentParser *config.Parser, intoCfg interface{}) error
> }
> 
> type UnmarshalFunc func(componentParser *config.Parser, intoCfg interface{}) error
> ```
> 
> @mx-psi you may need #2808 before in order to avoid circular deps.

Yep, made the changes for it. Thanks.

File: receiver/hostmetricsreceiver/config.go [hunk 7]
```diff
@@ -24,3 +25,10 @@ type Config struct {
 	scraperhelper.ScraperControllerSettings `mapstructure:",squash"`
 	Scrapers                                map[string]internal.Config `mapstructure:"-"`
 }
+
+var _ config.CustomConfigOptions = (*Config)(nil)
+
+// Validate implements the custom validation check on configuration
```
Comment (mx-psi) [inline receiver/hostmetricsreceiver/config.go:31]: nit: all methods 'implement' something, we could rephrase this to something like
```
// Validate checks the receiver configuration is valid
```

File: config/receiver.go [hunk 8]
```diff
@@ -16,10 +16,30 @@ package config
...
 	NamedEntity
+	CustomConfigOptions
 }
 
+// CustomConfigOptions defines the interfaces for configuration customization options
+// Different custom interfaces can be added like CustomConfigValidator, CustomConfigUnmarshaller, etc.
+type CustomConfigOptions interface {
+	CustomConfigValidator
+	// CustomConfigUnmarshaller
+}
```
Comment (mx-psi) [inline config/receiver.go:30]: From the discussion above we only want validation to be required since custom unmarshaling is more of an edge case so I would remove the references for now

```suggestion
// Different custom interfaces can be added like CustomConfigValidator, etc.
type CustomConfigOptions interface {
	CustomConfigValidator
}
```

**Review by mx-psi — COMMENTED** [review]
The design makes sense to me and I would say we can apply it to all component types (if Bogdan gives his 👍).

File: config/receiver.go (hunk 8 above)
Comment (mxiamxia) [reply config/receiver.go]: agree. updated

File: config/config_test.go [hunk 9]
```diff
@@ -16,11 +16,23 @@ package config
 
 import (
 	"errors"
+	"fmt"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 )
 
+type NopConfig struct {
```
Comment (bogdandrutu) [inline config/config_test.go:25]: No need to be public.

**Review by bogdandrutu — APPROVED** [review]
Let's merge as first draft, I will followup with some cleanups.

File: config/config_test.go (hunk 9 above)
Comment (mxiamxia) [reply config/config_test.go:25]: I'll move 

> Let's merge as first draft, I will followup with some cleanups.

@bogdandrutu , I am going to expand the same checks to all the other type of components in the next PR. Sounds good to you?
